### PR TITLE
#820 Make it a bit easier to run config functions without going through galsim executable

### PR DESCRIPTION
**Pull request by rmjarvis**
@esheldon reported on issue #820 that when running config functions by hand within python, e.g. `galsim.config.BuildGSObject(config, 'gal')`, it wasn't allowing the rng to be automatic for things that have `_takes_rng = True`.

This is normally not a problem when running config files through the `galsim` executable, since prior to that call, an appropriate rng will have bee constructed.  However, I think it does make sense to allow users to call functions such as this without requiring them to manually make the rng and put it in `config['rng']`.  So that's done here.

However, I also took this opportunity to go through the missing config coverage (which was a lot) and add unit tests to cover the things that weren't being covered by anything else yet.  Along the way, I found and fixed a number of bugs in edge case scenarios and made a few other changes to make it easier to call the config functions manually.  Here is a complete list of changes:
- Setting config['rng'] is no longer required when manually running commands like `galsim.config.BuildGSObject`.
- Changed the name of galsim.config.CalculateNoiseVar to the slightly more verbose, but also more readable CalculateNoiseVariance.
- Allow PoissonNoise without any sky level, in which case only the shot noise in the signal photons contribute to the noise.  Likewise for CCDNoise.
- Let 'None' in the config file mean `None`.
- Require 'max_extra_noise' to be explicitly set for photon shooting if you want it rather than have it default to 0.01.
- Added `--except_abort` option to `galsim` executable to abort execution if a file has an error, rather than just reporting the exception and continuing on (which is still the default behavior).
- Added optional probability parameter 'p' for Random bool values.
- Allow `drawImage` to take a non-uniform WCS when default constructing the image from either `nx`,`ny` or `bounds`.  (Note: this one is not a config change -- it was just a restriction that we had in `drawImage` that isn't actually necessary.)

Also a couple other functionality-neutral changes you'll notice here:
- When people used galsim.utilities for things, there was a mix of using the full `galsim.utilities.blah` vs `from . import utilities` ... `utilities.blah`.  It seemed that the former was more common, so I switched everything to use that style for consistency.
- Josh pointed out during the hack week that `_absdet` in SBTransform` wasn't actually correct.  It's really the total flux scaling.  And there was another parameter called `_fluxScaling`, which is really the amplitude scaling.  So I renamed these `_fluxScaling` and `_ampScaling` respectively to be clearer.  And I kept `_absdet` to actually mean |det(m)|.
- I also made SBConvolve and SBTransform delay some calculations until they are actually needed.  This won't usually be a performance change, since you usually do need them eventually when you draw the object, but for my `test_reject` function that builds a lot of objects and never gets around to drawing them, it dropped the running time from 7.5 seconds to a little over 2 seconds.

### CODE REVIEW

**Comment by rmandelb** [comment]
I'm getting some output from the unit tests on my Mac:
```
Traceback (most recent call last):
  File "/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/site-packages/galsim/config/process.py", line 822, in worker
    ps = pstats.Stats(pr, stream=s).sort_stats(sortby).reverse_order()
  File "/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/pstats.py", line 233, in sort_stats
    sort_tuple = sort_tuple + sort_arg_defs[word][0]
KeyError: 'tottime'
Process Process-4:
```

and on the Linux cluster, something completely different:
```
ERROR: Test the options in config to whiten images
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/python27/lib/python2.7/site-packages/nose-1.3.0-py2.7.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/rmandelb/git/GalSim/tests/galsim_test_helpers.py", line 559, in f2
    result = f(*args, **kwargs)
  File "/home/rmandelb/git/GalSim/tests/test_config_noise.py", line 661, in test_whiten
    im3d, cv3d = galsim.config.BuildStamp(config2)
  File "/home/rmandelb/software/lib/python2.7/site-packages/galsim/config/stamp.py", line 405, in BuildStamp
    im, current_var = builder.addNoise(stamp,config,im,skip,current_var,logger)
  File "/home/rmandelb/software/lib/python2.7/site-packages/galsim/config/stamp.py", line 876, in addNoise
    current_var = galsim.config.AddNoise(base,image,current_var,logger)
  File "/home/rmandelb/software/lib/python2.7/site-packages/galsim/config/noise.py", line 88, in AddNoise
    var = builder.addNoise(noise, config, im, rng, current_var, draw_method, logger)
  File "/home/rmandelb/software/lib/python2.7/site-packages/galsim/config/noise.py", line 326, in addNoise
    "Whitening already added more noise than the requested Poisson noise.")
RuntimeError: Whitening already added more noise than the requested Poisson noise.
-------------------- >> begin captured stdout << ---------------------
From whiten, current_var =  9.81378306774e-05
From BuildStamp, current_var =  9.81378306774e-05
```

**Comment by rmjarvis** [comment]
> RuntimeError: Whitening already added more noise than the requested Poisson noise.

Travis found that one too, so I was already working on it.  I'm not actually sure what the error was, but when I cleaned up that bit of code slightly to not use quite so many temporary numpy arrays, the error went away.  Weird.  cf.1b6060d

> KeyError: 'tottime'

This seems to indicate that you have an old version of pstats.  AFAICT, the name for this used to be 'time' and they switched it to 'tottime', but 'time' is still a valid alias.  So assuming this alias will remain valid going forward, I just switched to that.  cf. 7f7d909c5c

**Comment by rmandelb** [comment]
I confirm these are both fixed, thanks.  I plan to go over the code sometime before Thanksgiving (but likely not until Weds or so).

File: tests/test_config_value.py [hunk 1]
```diff
@@ -980,6 +1100,78 @@ def test_shear_value():
...
+                                 galsim.Shear)
+        config['world_pos'] = galsim.PositionD(6,8)
+        # Still raise an error because no redshift
+        np.testing.assert_raises(ValueError, galsim.config.ParseValue, config, 'nfw', config,
+                                 galsim.Shear)
+        # With this, it should work.
+        config['gal'] = { 'redshift' : 1.3 }
+    except ImportError:
+        pass
+    nfw_halo = galsim.NFWHalo(mass=1.e14, conc=4, redshift=0.3)
```
Comment (rmandelb) [inline tests/test_config_value.py]: Sorry to pick on this random example, but there are lots of magic numbers that need to be written identically between the config and the direct cases, so this is another test routine that might benefit from defining them as variables.  This is true in quite a few of the new tests.

File: CHANGELOG.md [hunk 2]
```diff
@@ -106,3 +108,17 @@ New config features
 
 - Output slightly more information about the COSMOSCatalog() (if any) being used
   as the basis of simulations, at the default verbosity level. (#804)
+- Changed the name of galsim.config.CalculateNoiseVar to the slightly more
+  verbose, but also more readable CalculateNoiseVariance. (#820)
+- Setting config['rng'] is no longer required when manually running commands
+  like `galsim.config.BuildGSObject`.  If you do not care about deterministic
+  pseudo-rngs, then it will just use /dev/urandom for you. (#820)
```
Comment (rmjarvis) [inline CHANGELOG.md]: Sure.  I can add that.  Although, it would require that you are running things with a logger, since that's where config warnings are emitted, and I might guess that most people who neglect to specify the rng might also neglect to pass in a logger.

File: CHANGELOG.md [hunk 3]
```diff
@@ -106,3 +108,17 @@ New config features
...
+- Allow PoissonNoise without any sky level, in which case only the shot noise
+  in the signal photons contribute to the noise.  Likewise for CCDNoise. (#820)
```
Comment (rmjarvis) [inline CHANGELOG.md]: I agree, and that's why it wasn't there before.  But it seemed like a silly restriction to not allow using PoissonNoise without a sky.  It seems perfectly legitimate to investigate galaxies where the only noise is from their own photons.

File: galsim/config/noise.py [hunk 4]
```diff
@@ -506,9 +490,9 @@ def addNoiseVariance(self, config, base, im, include_obj_var, logger):
             # The current image at this point should be the noise-free, sky-free image,
             # which is the object variance in each pixel.
             if gain != 1.0:
-                im += base['current_image']/gain
+                im += base['current_noise_image']/gain
```
Comment (rmjarvis) [inline galsim/config/noise.py]: It's just an efficiency.  Dividing by 1 is not a trivial operation on a image.  So it's better to check first and skip the calculation if it won't actually do anything.

File: tests/test_config_image.py [hunk 5]
```diff
@@ -31,9 +33,612 @@
...
+    im4a = gal.drawImage(scale=1, method='phot', max_extra_noise=5, rng=ud, poisson_flux=True)
+    im4a.addNoise(galsim.GaussianNoise(sigma=math.sqrt(50), rng=ud))
+    im4b = galsim.config.BuildImage(config)
+    np.testing.assert_array_equal(im4b.array, im4a.array)
+
+@timer
+def test_reject():
+    """Test various ways that objects can be rejected.
+    """
+    # Make a custom function for rejecting COSMOSCatalog objects that use Sersics with n > 2.
```
Comment (rmjarvis) [inline tests/test_config_image.py]: Do you think this is more helpful than the example in cgc.yaml?

Comment (rmandelb) [inline tests/test_config_image.py]: I don't think it's _more_ helpful.  I just think it's another helpful thing.  They showcase different functionality (i.e., the one in cgc.yaml is based on properties of the drawn image whereas the one in this test lets you really dig into the COSMOSCatalog, and I don't find that trivial/obvious, personally).

File: galsim/config/noise.py (hunk 4 above)
Comment (rmandelb) [inline galsim/config/noise.py]: OK, I wouldn't have guessed that this mattered.

File: galsim/config/extra.py [hunk 6]
```diff
@@ -164,71 +162,70 @@ def WriteExtraOutputs(config, main_data, logger=None):
...
     config['index_key'] = 'file_num'
-    if 'output' in config:
-        output = config['output']
-        if 'retry_io' in output:
-            ntries = galsim.config.ParseValue(config['output'],'retry_io',config,int)[0]
-            # This is how many retries.  Do at least 1, so ntries is 1 more than this.
-            ntries = ntries + 1
-        else:
-            ntries = 1
+    output = config['output']
```
Comment (rmandelb) [inline galsim/config/extra.py]: Cool.

File: CHANGELOG.md (hunk 2 above)
Comment (rmandelb) [inline CHANGELOG.md]: Oh.  Well, that's not terribly satisfying.

File: examples/demo7.yaml [hunk 7]
```diff
@@ -191,6 +191,17 @@ stamp :
...
+    # For bright objects, it can be quite slow to shoot all the photons from the object.
+    # Shooting fewer photons with flux > 1 each will be faster, but will add some extra
+    # Poisson noise due to the smaller number of photons.
+    # The max_extra_noise parameter indicates how much extra noise per pixel we are
+    # willing to tolerate as a fraction of the noise that we will be adding from the
+    # sky (cf. the image.noise field).  In this case, we allow an extra 1% larger variance,
+    # but only for photon
+    max_extra_noise :
+        type: List
+        items: [ None, 0.01 ]
```
Comment (rmjarvis) [inline examples/demo7.yaml]: That was what I was starting to write in that sentence that I didn't finish and apparently never came back to.

File: galsim/config/noise.py (hunk 4 above)
Comment (rmjarvis) [inline galsim/config/noise.py]: It likely would be a very minor inefficiency, you're right.  But these little things can add up.  :)

File: tests/test_config_image.py (hunk 5 above)
Comment (rmjarvis) [inline tests/test_config_image.py]: I added something [here](https://github.com/GalSim-developers/GalSim/wiki/Config-Stamp#stamp-types).  The link is to the master version, which doesn't actually have this code yet, but it will once this is merged.

File: tests/test_config_value.py (hunk 1 above)
Comment (rmjarvis) [inline tests/test_config_value.py]: OK. I did these two, but I didn't try to comprehensively go through the tests to look for examples where doing this would be helpful.  If you notice any others that you think are egregious, let me know.  (Personally, I don't much mind magic numbers in tests -- in the main code is where I try to avoid them.)

**Comment by rmjarvis** [comment]
Thanks Rachel.  I think I addressed everything.

**Comment by rmjarvis** [comment]
Any objection to merging this?

**Comment by jmeyers314** [comment]
>Any objection to merging this?

I bet you were waiting to hear from me on this...  Let me give it a quick once-over.

**Comment by rmjarvis** [comment]
:)  Yep.  Thanks.

**Comment by jmeyers314** [comment]
It was a quick scan, but looks good to me.  If nothing else, I learned of the existence of `pragma: no branch` and `RealGalaxyOriginal` :).

**Comment by rmjarvis** [comment]
Great.  Thanks, Josh.
